components/test_raftstore-v2/src/util.rs

```diff
@@ -0,0 +1,153 @@
+// Copyright 2022 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::{fmt::Write, sync::Arc, time::Duration};
+
+use encryption_export::{data_key_manager_from_config, DataKeyManager};
+use engine_rocks::{RocksEngine, RocksStatistics};
+use engine_test::raft::RaftTestEngine;
+use engine_traits::{TabletRegistry, CF_DEFAULT};
+use file_system::IoRateLimiter;
+use rand::RngCore;
+use server::server2::ConfiguredRaftEngine;
+use tempfile::TempDir;
+use test_raftstore::{new_put_cf_cmd, Config};
+use tikv::{server::KvEngineFactoryBuilder, storage::config::EngineType};
+use tikv_util::{config::ReadableDuration, worker::LazyWorker};
+
+use crate::{bootstrap_store, cluster::Cluster, ClusterType, Simulator};
+
+pub fn create_test_engine(
+    // TODO: pass it in for all cases.
+    id: Option<(u64, u64)>,
+    limiter: Option<Arc<IoRateLimiter>>,
+    cfg: &Config,
+) -> (
+    TabletRegistry<RocksEngine>,
+    RaftTestEngine,
+    Option<Arc<DataKeyManager>>,
+    TempDir,
+    LazyWorker<String>,
+    Arc<RocksStatistics>,
+    Option<Arc<RocksStatistics>>,
+) {
+    let dir = test_util::temp_dir("test_cluster", cfg.prefer_mem);
+    let mut cfg = cfg.clone();
+    cfg.storage.data_dir = dir.path().to_str().unwrap().to_string();
+    cfg.raft_store.raftdb_path = cfg.infer_raft_db_path(None).unwrap();
+    cfg.raft_engine.mut_config().dir = cfg.infer_raft_engine_path(None).unwrap();
+    let key_manager =
+        data_key_manager_from_config(&cfg.security.encryption, dir.path().to_str().unwrap())
+            .unwrap()
+            .map(Arc::new);
+    let cache = cfg
+        .storage
+        .block_cache
+        .build_shared_cache(EngineType::RaftKv2);
+    let env = cfg
+        .build_shared_rocks_env(key_manager.clone(), limiter)
+        .unwrap();
+
+    let sst_worker = LazyWorker::new("sst-recovery");
+    let scheduler = sst_worker.scheduler();
+
+    let (raft_engine, raft_statistics) = RaftTestEngine::build(&cfg, &env, &key_manager, &cache);
+
+    if let Some((cluster_id, store_id)) = id {
+        assert_ne!(store_id, 0);
+        bootstrap_store(&raft_engine, cluster_id, store_id).unwrap();
+    }
+
+    let builder =
+        KvEngineFactoryBuilder::new(env, &cfg.tikv, cache).sst_recovery_sender(Some(scheduler));
+
+    let factory = Box::new(builder.build());
+    let rocks_statistics = factory.rocks_statistics();
+    let reg = TabletRegistry::new(factory, dir.path().join("tablet")).unwrap();
+
+    (
+        reg,
+        raft_engine,
+        key_manager,
+        dir,
+        sst_worker,
+        rocks_statistics,
+        raft_statistics,
+    )
+}
+
+/// Keep putting random kvs until specified size limit is reached.
+pub fn put_till_size(
+    cluster: ClusterType<'_>,
```

Comment:
So far I see there're three functions using `ClusterType`, will there be more? And approximately how many more?

Reply:
There will  be more functions using it. Actually,  the functions like `fn func1(cluster: &mut Cluster)` cannot be handled when there are multiple kinds of Cluster types (v1 and v2). ClusterType can be used to hide the type and calls the cluster's related functions based on the Cluster type.

Reply:
To unify different clusters under the same trait makes proc macro not as useful. I prefer we use `let put_till_size = |c| {...}` instead so there's no need to abstract different cluster types. I skimmed through some of the functions that used `&mut Cluster`, majority of them are called only once. So the refactoring should be easy. For functions that are used multiple times (in different tests), maybe merge them into one single test case.

Reply:
put_till_size and put_cf_till_size are used in different files. Using let put_till_size = |c| {...} means we will copy it in multiple places?

Reply:
For this particular case, there're two methods to go around it. (1) make `put_till_size` a cluster method (instead of a standalone util function) (2) import `test_raftstore::util::*` and `test_raftstore_v2::util::*`. I vaguely remember rust will choose the correct implementation if there're many with the same name. If rust fails to compile, you can move the import into proc macro. i.e. only import the util::* for the corresponding crate specified in proc macro.

Reply:
I will have a try, thanks.